cs3api4lab/api/cs3_share_api.py

```diff
@@ -0,0 +1,279 @@
+"""
+cs3_share_api.py
+
+CS3 Share API for the JupyterLab Extension
+
+Authors:
+"""
+
+import re
+import cs3.sharing.collaboration.v1beta1.collaboration_api_pb2 as sharing
+import cs3.sharing.collaboration.v1beta1.resources_pb2 as sharing_res
+import cs3.storage.provider.v1beta1.provider_api_pb2 as storage_provider
+import cs3.storage.provider.v1beta1.resources_pb2 as storage_resources
+import cs3.identity.user.v1beta1.resources_pb2 as identity_res
+import cs3.rpc.v1beta1.code_pb2 as cs3_code
+import random
+from cs3api4lab.auth.authenticator import Authenticator
+from cs3api4lab.api.cs3_file_api import Cs3FileApi
+from cs3api4lab.api.file_utils import FileUtils
+from cs3api4lab.common.strings import *
+from cs3api4lab.config.config_manager import Cs3ConfigManager
+import cs3.gateway.v1beta1.gateway_api_pb2_grpc as grpc_gateway
+from cs3api4lab.auth.channel_connector import ChannelConnector
+
+
+class Cs3ShareApi:
+    gateway_stub = None
+    log = None
+    auth = None
+    config = {}
+    file_api = None
+
+    def __init__(self, log):
+        self.config = Cs3ConfigManager().config
+        self.auth = Authenticator()
+        self.file_api = Cs3FileApi(log)
+        self.gateway_stub = grpc_gateway.GatewayAPIStub(ChannelConnector().channel)
+        self.log = log
+        return
+
+    def create(self, endpoint, fileid, grantee, idp, role=Role.VIEWER, grantee_type=Grantee.USER):
+        if grantee is None:
+            raise Exception("Grantee was not provided")
+        share_permissions = self._get_share_permissions(role)
+        grantee_type_enum = self._get_grantee_type(grantee_type)
+        share_grant = self._get_share_grant(grantee_type_enum, share_permissions, idp, grantee)
+        resource_info = self._get_resource_info(endpoint, fileid)
+        share_request = sharing.CreateShareRequest(resource_info=resource_info, grant=share_grant)
+        token = self.get_token()
+        share_response = self.gateway_stub.CreateShare(request=share_request,
+                                                       metadata=[('x-access-token', token)])
+        self._check_response_code(share_response)
+        self.log.info("Share created")
+        self.log.info(share_response)
+        return self._map_given_share(share_response.share)
+
+    def list(self):
+        # todo filters
+        list_request = sharing.ListSharesRequest()
+        list_response = self.gateway_stub.ListShares(request=list_request,
+                                                     metadata=[('x-access-token', self.get_token())])
+        self._check_response_code(list_response)
+        self.log.info("List shares response for user: " + self.config['client_id'])
+        self.log.info(list_request)
+        return self._map_given_shares(list_response)
```

Comment:
I would return a list of shares that map a fileid/etc -> list of people it is shared with.
It makes more sense to have a list of all the unique shares available, instead of returning the repeated values in case it's shared with multiple users.
So I recommend we group by ID/path/some identifier.